src/Libraries/Microsoft.Extensions.Telemetry/Sampling/SamplingLoggerBuilderExtensions.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Logging;
+using Microsoft.Shared.DiagnosticIds;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.Diagnostics.Sampling;
+
+/// <summary>
+/// Extensions for configuring logging sampling.
+/// </summary>
+[Experimental(diagnosticId: DiagnosticIds.Experiments.Telemetry, UrlFormat = DiagnosticIds.UrlFormat)]
+public static class SamplingLoggerBuilderExtensions
+{
+    /// <summary>
+    /// Adds Trace-based logging sampling to the logging infrastructure. Sampling decisions
+    /// for logs match exactly the sampling decisions for the underlying <see cref="System.Diagnostics.Activity"/>.
+    /// </summary>
+    /// <param name="builder">The dependency injection container to add logging to.</param>
+    /// <returns>The value of <paramref name="builder"/>.</returns>
+    /// <remarks>Please configure Tracing Sampling separately as part of OpenTelemetry .NET.</remarks>
+    /// <exception cref="ArgumentNullException"><paramref name="builder"/> is <see langword="null"/>.</exception>
+    public static ILoggingBuilder AddTraceBasedSampling(this ILoggingBuilder builder)
+    {
+        _ = Throw.IfNull(builder);
+
+        return builder.AddSampler<TraceBasedSampler>();
+    }
+
+    /// <summary>
+    /// Adds Ratio-based sampler to the logging infrastructure. Matched logs will be sampled
+    /// according to the provided <paramref name="probability"/>.
+    /// Higher the probability value, higher is the probability of a given log record to be sampled in.
+    /// </summary>
+    /// <param name="builder">The dependency injection container to add logging to.</param>
+    /// <param name="probability">Probability from 0.0 to 1.0.</param>
+    /// <param name="level">The log level (and below) to apply the sampler to.</param>
+    /// <param name="category">The log category to apply the sampler to.</param>
+    /// <param name="eventId">The event ID to apply the sampler to.</param>
+    /// <returns>The value of <paramref name="builder"/>.</returns>
+    /// <exception cref="ArgumentNullException"><paramref name="builder"/> is <see langword="null"/>.</exception>
+    public static ILoggingBuilder AddRatioBasedSampler(
+        this ILoggingBuilder builder,
+        double probability,
+        LogLevel? level = null,
+        string? category = null,
```

Comment:
I think we may want to make level be the only optional parameter and get rid of category/eventId. It feels very odd to say the sampler should only filter one category or one eventId given that users are limited to one sampler in total. It feels simpler to have them write their own sampler as soon as they need any category or eventId logic in their sampling. Alternately we could implement a more complicated RatioBasedSampler that allows specifying a list of different categories but I'd keep it simple until we've seen a clear need.

Reply:
Updated